forge/src/runner.rs

```diff
@@ -210,251 +186,234 @@ impl<'a, DB: DatabaseRef + Send + Sync> ContractRunner<'a, DB> {
     pub fn run_tests(
         &mut self,
         filter: &impl TestFilter,
         fuzzer: Option<TestRunner>,
-        known_contracts: Option<&BTreeMap<String, (Abi, Vec<u8>)>>,
     ) -> Result<BTreeMap<String, TestResult>> {
         tracing::info!("starting tests");
         let start = Instant::now();
         let needs_setup = self.contract.functions().any(|func| func.name == "setUp");
-        let (unit_tests, fuzz_tests): (Vec<_>, Vec<_>) = self
-            .contract
-            .functions()
-            .into_iter()
-            .filter(|func| func.name.starts_with("test"))
-            .filter(|func| filter.matches_test(&func.name))
-            .partition(|func| func.inputs.is_empty());
 
-        let (addr, init_logs, setup_failed, reason) = self.deploy(needs_setup)?;
-        if setup_failed {
+        let setup = self.setup(needs_setup)?;
+        if setup.setup_failed {
             // The setup failed, so we return a single test result for `setUp`
             return Ok([(
                 "setUp()".to_string(),
                 TestResult {
                     success: false,
-                    reason,
-                    gas_used: 0,
+                    reason: setup.reason,
                     counterexample: None,
-                    logs: init_logs,
+                    logs: setup.logs,
                     kind: TestKind::Standard(0),
-                    traces: None,
-                    identified_contracts: None,
+                    traces: setup.traces,
+                    // TODO
                     debug_calls: None,
-                    labeled_addresses: Default::default(),
+                    labeled_addresses: setup.labeled_addresses,
                 },
             )]
             .into())
         }
 
-        // Run all unit tests
-        let mut test_results = unit_tests
+        // Collect valid test functions
+        let tests: Vec<_> = self
+            .contract
+            .functions()
+            .into_iter()
+            .filter(|func| func.name.starts_with("test") && filter.matches_test(&func.name))
+            .map(|func| (func, func.name.starts_with("testFail")))
+            .collect();
+
+        let test_results = tests
             .par_iter()
-            .map(|func| {
-                let result = self.run_test(func, known_contracts, addr, init_logs.clone())?;
-                Ok((func.signature(), result))
+            .filter_map(|(func, should_fail)| {
+                let result = if func.inputs.is_empty() {
+                    Some(self.run_test(func, *should_fail, setup.clone()))
+                } else {
+                    fuzzer.as_ref().map(|fuzzer| {
+                        self.run_fuzz_test(func, *should_fail, fuzzer.clone(), setup.clone())
+                    })
+                };
+
+                result.map(|result| Ok((func.signature(), result?)))
             })
             .collect::<Result<BTreeMap<_, _>>>()?;
```

Comment:
Previously we ran unit tests in parallel, waited, and then ran fuzz tests in parallel, but now we just run all of them in parallel